qualtricssurvey/models.py

```diff
@@ -483,11 +564,131 @@ def should_show_simulation_exists(self):
         Return True/False to indicate whether to show the "Show Qualtrics Survey Meta Information" information.
         """
         return self.show_meta_information
+              
+    def max_score(self):
+        """
+        Return the weight of the problem. This method is in the staff debug information
+        """
+        return self.weight
+
+    def get_is_graded(self):
+        try:
+            survey_status = SurveyStatus.objects.get(usage_key=self.location, user_id=self.xmodule_runtime.user_id).status
+            if (survey_status == "Complete"):
+                is_graded = "Graded"
+                return is_graded
+        except:
+            pass
+
+        is_graded = "Ungraded"   
+        return is_graded
+
+    def get_earned_score(self):
+        return self.score.raw_earned
+
+    def set_earned_score(self):
+        self.earned_score = self.weight
+
+    def publish_grade(self):
+        grade_dict = {
+            'value': self.score.raw_earned,
+            'max_value': self.score.raw_possible,
+        }
+        self.runtime.publish(self, "grade", grade_dict)
+
+    @XBlock.json_handler
+    def get_survey_status(self, data, suffix=''):
+        try:
+            survey_status = SurveyStatus.objects.get(usage_key=self.location, user_id=self.xmodule_runtime.user_id).status
+                
+        except:
+            survey_status = "Incomplete"
+            grade = 0
+
+        # Prevents dividing by zero when computing weighted score for unweighted survey
+        if (self.score.raw_possible != 0) :
+            earned_score = (self.score.raw_earned/self.score.raw_possible) * self.weight
+        
+        else: 
+            earned_score = 0
 
+
+        return {'survey_status': survey_status, 'max_score': self.weight, 'earned_score': earned_score}
+        
+
+    @QualtricsHandlersMixin.x_www_form_handler
+    def end_survey(self, data, suffix=''):  # pylint: disable=unused-argument
+        """
+        Called upon completion of the survey
+        """
+        
+        survey_id = data.get("SurveyID")
+        response_id = data.get("ResponseID")
+        url = "https://clemson.qualtrics.com/API/v3/surveys/{}/responses/{}".format(survey_id, response_id)
+
+        payload={}
+        headers = {
+        'X-API-TOKEN': 'bJjfXqGYjXqp0triy3dnRmwD1vZ6lXFeAw41GTLW'
+        }
+
+        response_survey = requests.request("GET", url, headers=headers, data=payload)
+
+        data_response_survey = response_survey.json()
+        
+        response = {
+            "Message": "Data processed from the Qualtrics Event Subscription API postback `surveyengine.completedResponse` event."
+        }
+        status = data.get("Status")
+
+        if status == "Complete":
+            result = data_response_survey["result"]
+            values = result["values"]
+            real_user = user_by_anonymous_id(values["anonymous_user_id"])
+            self.system.rebind_noauth_module_to_user(self, real_user)
```

Comment:
@bsande6 Can you add a comment here explaining why we're having to do this rebind? If someone looked at this later they could get confused.